connect/runtime/src/main/java/org/apache/kafka/connect/runtime/WorkerTask.java

```diff
@@ -133,22 +138,69 @@ private void doRun() {
         }
     }
 
+    private void onShutdown() {
+        synchronized (this) {
+            stopping.set(true);
+            // if we were cancelled, skip the status update since the task may have already been
+            // started somewhere else
+            if (!cancelled.get())
+                statusListener.onShutdown(id);
+        }
+    }
+
+    private void onFailure(Throwable t) {
+        synchronized (this) {
+            stopping.set(true);
+            // if we were cancelled, skip the status update since the task may have already been
+            // started somewhere else
+            if (!cancelled.get())
+                statusListener.onFailure(id, t);
+        }
+    }
+
     @Override
     public void run() {
         try {
             doRun();
-            if (!cancelled.get())
-                lifecycleListener.onShutdown(id);
+            onShutdown();
         } catch (Throwable t) {
-            if (!cancelled.get())
-                lifecycleListener.onFailure(id, t);
+            onFailure(t);
 
             if (t instanceof Error)
                 throw t;
         } finally {
-            running.set(false);
             shutdownLatch.countDown();
         }
     }
 
+    public boolean isPaused() {
+        return this.targetState.get() == TargetState.PAUSED;
+    }
+
+    protected void awaitUnpause(long timeoutMs) throws InterruptedException {
+        synchronized (this) {
+            if (targetState.get() != TargetState.PAUSED) return;
+            this.wait(timeoutMs);
```

Comment:
`Object.wait` may see spurious wakeups, if we want to properly handle the timeout we'd need a loop here w/ a `Time` object to be able to actually check whether the full time has elapsed.